File: server/realtime/ws.go
package realtime

import "github.com/gorilla/websocket"

// EventKind represents the type of an event or eventResponse that's sent over a
// Websocket connection.
type EventKind string

const (
	// LobbyInfo describes an event that the server sends when a client first
	// connects to a /:gameID URL. It communicates whether a game with the
	// provided ID already exists or not, as well as the other players that are
	// already in that game.
	LobbyInfo EventKind = "LOBBY_INFO"
	// NewPlayerID describes an event that the server sends to a client after
	// making a new Player object for them. Clients need to know what their
	// Player IDs are in order to include in the bodies of select events.
	NewPlayerID EventKind = "NEW_PLAYER_ID"
	// NewPlayerJoined describes an event that a client sends after it first
	// connects to a /:gameID URL and sets its display name. This gets
	// broadcasted to all other Players in that game lobby so that they may add
	// this new player to their red teams client-side.
	NewPlayerJoined EventKind = "NEW_PLAYER_JOINED"
	// ChangeDisplayName describes an event that the client sends when they
	// change their display name.
	ChangeDisplayName EventKind = "CHANGE_DISPLAY_NAME"
	// SomeoneElseChangeDisplayName describes an event that the server sends to
	// all clients in a game if one of the players changes their display name.
	SomeoneElseChangeDisplayName EventKind = "SOMEONE_ELSE_CHANGE_DISPLAY_NAME"
	// ChangeTeam describes an event that the client has swapped teams in a game
	// lobby.
	ChangeTeam EventKind = "CHANGE_TEAM"
	// SomeoneElseChangeTeam describes an event that some other Player in a game
	// lobby changed teams.
	SomeoneElseChangeTeam EventKind = "SOMEONE_ELSE_CHANGE_TEAM"
)

// event mirrors the structure of JSON messages that clients send to the server
// via a Websocket connection.
type event struct {
	Kind EventKind   `json:"kind"`
	Body interface{} `json:"body"`
}

// eventResponse mirrors the structure of JSON messages that the server sends in
// response to a client's Websocket message.
type eventResponse struct {
	Ok   bool        `json:"ok"`
	Kind EventKind   `json:"kind"`
	Body interface{} `json:"body"`
}

// broadcastMsg objects are stored in an Interactor's broadcast channel. They
// contain an event to be broadcasted to all Players in a Game except for the
// Player who originally sent the event.
type broadcastMsg struct {
	// ID of the Player who originally sent the event. This is so that Players
	// who send events that are broadcasted to everyone in a Game don't receive
	// their own event.
	originClientID string
	// The event to broadcast to all other Players in a Game.
	event *event
}

// ConstructAndSendEvent builds an event struct with the provided fields,
// marshals it to JSON, and sends it along the provided Websocket connection.
func ConstructAndSendEvent(conn *websocket.Conn, kind EventKind, body interface{}) {
	event := event{Kind: kind, Body: body}
	conn.WriteJSON(event)
}

// constructAndSendResponse builds an eventResponse struct with the provided
// fields, marshals it to JSON, and sends it along the provided Websocket
// connection. Body parameter may be nil.
func constructAndSendResponse(
	conn *websocket.Conn,
	kind EventKind,
	body interface{},
) {
	response := eventResponse{Ok: true, Kind: kind}
	if body != nil {
		response.Body = body
	}
	conn.WriteJSON(response)
}

// constructAndSendErr builds an eventResponse struct with the provided error,
// marshals it to JSON, and sends it along the provided Websocket connection.
func constructAndSendErr(conn *websocket.Conn, err error) {
	response := eventResponse{Ok: false, Body: err}
	conn.WriteJSON(response)
}

File: client/src/realtime/ws.ts
import { store, RootState } from '../store/index';
import { SERVER_URL } from '../constants';
import { setIsCreated, setIsJoined } from '../store/game/actions';
import { setUserID } from '../store/user/actions';

export enum EventKind {
  lobbyInfo = 'LOBBY_INFO',
  newPlayerID = 'NEW_PLAYER_ID',
  changeDisplayName = 'CHANGE_DISPLAY_NAME',
  notAnEvent = 'NOT_AN_EVENT',
}

// Event mirrors the structure of JSON messages that the server sends to clients
// via Websocket connections when something occurrs that all clients in a game
// need to be aware of.
interface Event {
  kind: EventKind;
  body: any;
}

// EventResponse mirrors the structure of JSON messages that the server sends to
// a client via a Websocket connection in response to an event that that client
// sent to the server.
interface EventResponse {
  ok: boolean;
  kind: EventKind;
  body?: any;
}

// constructAndSendEvent builds an Event with the provided fields, serializes it
// into a string, and sends it along the provided Websocket connection.
export function constructAndSendEvent(
  socket: WebSocket,
  kind: EventKind,
  body: any,
): void {
  const event: Event = { kind, body };
  socket.send(JSON.stringify(event));
}

// constructAndSendResponse builds an EventResponse with the provided fields,
// serializes it into a string, and sends it along the provided Websocket
// connection. Body parameter may be omitted.
export function constructAndSendResponse(
  socket: WebSocket,
  kind: EventKind,
  body?: any,
): void {
  const response: EventResponse = { ok: true, kind };
  if (body) {
    response.body = body;
  }
  socket.send(JSON.stringify(response));
}

// constructAndSendErr builds an EventResponse with the provided fields,
// serializes it into a string, and sends it along the provided Websocket
// connection. Body parameter may be omitted.
export function constructAndSendErr(
  socket: WebSocket,
  kind: EventKind,
  body?: any,
): void {
  const response: EventResponse = { ok: false, kind };
  if (body) {
    response.body = body;
  }
  socket.send(JSON.stringify(response));
}

// getEventKind makes sure that the provided event object is an Event, and
// returns the EventKind of that event. It's also the future home of logging or
// some kind of universal error handling logic.
export function getEventKind(event: any): EventKind {
  if (!('kind' in event)) {
    return EventKind.notAnEvent;
  }
  return event.kind;
}

// isResponseOK returns whether an event response object is carrying information
// about an something that went wrong. It's also the future home of logging or
// some kind of universal error handling logic for error responses that indicate
// there's a problem somewhere.
export function isResponseOK(response: EventResponse): boolean {
  return response.ok;
}

// getResponseBody returns the body of an EventResponse that indicated something
// was successful. If the provided EventResponse doesn't indicate that something
// was successful, then null is returned.
export function getResponseBody(response: EventResponse): any {
  if (!isResponseOK(response)) {
    return null;
  }
  return response.body;
}
// getResponseErr returns the body of an EventResponse that indicated something
// went wrong somewhere. If the provided EventResponse doesn't indicate that
// something went wrong, then null is returned.
export function getResponseErr(response: EventResponse): any {
  if (isResponseOK(response)) {
    return null;
  }
  return response.body;
}

// notifyOfUnrecognizedEvent is called when we receive a Websocket event from
// the server with an EventKind that we don't recongize.
function notifyOfUnrecognizedEvent(event: Event) {
  // TODO: better error handling.
  alert('Unrecognized Websocket event from the server. Check the console.');
  console.log(event);
}

// handleMsgFromServer parses a Websocket message from the server, and acts
// appropriately depending on that event's kind and the client's state (have we
// joined a game? Is the game we're trying to join created already?)
function handleMsgFromServer(msg: MessageEvent) {
  const globalState = store.getState();
  const event: Event = JSON.parse(msg.data);

  if (globalState.game.isJoined) {
    switch (event.kind) {
      default:
        notifyOfUnrecognizedEvent(event);
        break;
    }
  } else {
    switch (event.kind) {
      case EventKind.newPlayerID:
        // When a client first connects to a game, the server creates a new
        // Player object for them with a UUID. The client needs to include this
        // UUID in the body of select events that it sends.
        store.dispatch(setUserID(event.body.id));
        break;
      case EventKind.lobbyInfo:
        // When a client first visits a /:gameID URL and a Websocket connection
        // with the server is established, the server will send down a lobbyInfo
        // event to tell the client whether a game with the provided ID already
        // exists or not.
        store.dispatch(setIsCreated(event.body.isCreated));
        store.dispatch(setIsJoined(false));
        break;
      default:
        notifyOfUnrecognizedEvent(event);
        break;
    }
  }
}

// establishWSConnection attempts to establish a persistent Websocket connection
// with the server.
export function establishWSConnection(gameID: string): WebSocket {
  const protocol = window.location.protocol === 'https:' ? 'wss://' : 'ws://';
  const socketURL = `${protocol}${SERVER_URL}/ws?gameID=${gameID}`;
  const socket = new WebSocket(socketURL);

  socket.onmessage = (msg) => handleMsgFromServer(msg);
  // TODO: better error handling.
  socket.onerror = (event) => {
    alert(
      'Something went wrong with the Websocket connection to the server. Check the console.',
    );
    console.log(event);
    socket.close(1006);
  };

  return socket;
}
